apiserver/apiservermetrics.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package apiserver
+
+import (
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	apiserverMetricsNamespace = "juju_apiserver"
+)
+
+// ServerMetricsSource implementations provide apiserver metrics.
+type ServerMetricsSource interface {
+	ConnectionCount() int64
+	ConcurrentLoginAttempts() int64
+	ConnectionPauseTime() time.Duration
+	ConnectionRate() int64
+}
+
+// Collector is a prometheus.Collector that collects metrics based
+// on apiserver status.
+type Collector struct {
+	src ServerMetricsSource
+
+	connectionCountGauge     prometheus.Gauge
+	connectionRateGauge      prometheus.Gauge
+	connectionPauseTimeGauge prometheus.Gauge
+	concurrentLoginsGauge    prometheus.Gauge
+}
+
+// NewMetricsCollector returns a new Collector.
+func NewMetricsCollector(src ServerMetricsSource) *Collector {
+	return &Collector{
+		src: src,
+		connectionCountGauge: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: apiserverMetricsNamespace,
+			Name:      "connection_count",
+			Help:      "Current number of apiserver connections",
+		}),
+		connectionRateGauge: prometheus.NewGauge(prometheus.GaugeOpts{
```

Comment:
if we also exported a counter metric for connections, users could use prometheus's builtin rate() function for computing the per-second rate of change...